## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

data = pd.read_csv('Resources/client_dataset.csv')

data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
data.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
data.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
data.dtypes



first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [6]:
# What three item categories had the most entries?
top_3_categories = data["category"].value_counts().head(3)
print(top_3_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [13]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = data["category"].value_counts().index[0]
subcategory_setup = data[data["category"] == top_category]
top_subcategory = subcategory_setup["subcategory"].value_counts().index[0]

common_categories = data["category"].value_counts().idxmax()
most_common_category = data[data["category"] == common_categories]
most_common_subcategory = most_common_category["subcategory"].value_counts().max()

top_subcategory, most_common_subcategory


('bathroom supplies', 6424)

In [14]:
# Which five clients had the most entries in the data?
top_5_client_id_entries = data["client_id"].value_counts().head(5)
top_5_client_id_entries


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [19]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = top_5_client_id_entries.index.tolist()
top_5_client_ids


[33615, 66037, 46820, 38378, 24741]

In [22]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_5_client_entries = data["client_id"].value_counts().idxmax()
cleaned_clients = data[data["client_id"] == top_5_client_entries]
top_client_qty = cleaned_clients["qty"].sum()
top_client_qty


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [23]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
data["subtotal_calculation"] = data["unit_price"] * data["qty"]
price_quantity_subtotal = data[["unit_price", "qty", "subtotal_calculation"]]
price_quantity_subtotal.head(2)


unit_price  qty  subtotal_calculation
0     1096.80  105             115164.00
1       24.95   21                523.95

In [189]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def shipping_price(unit_weight):
    if unit_weight > 50:
        return 7 * unit_weight
    else:
        return 10 * unit_weight
    
data["total_weight"] = data["unit_weight"] * data["qty"]

data["shipping_price"] = data["total_weight"].apply(shipping_price) 
data["shipping_price"] = data["shipping_price"].round(2)

shippingtotal = data[["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]]

shippingtotal.head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [188]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
data["line_price"] = data["shipping_price"] + data["subtotal_calculation"] 

sales_tax = 0.0925

data["line_price"] = data["line_price"] * (1 + sales_tax)
data["line_price"] = data["line_price"].round(2)

data.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   subtotal_calculation  total_weight  shipping_price  line_price  line_cost  \
0             115164.00        787.50         5512.50   131839.08   85597.05   
1                523.95         31.29          312.90      914.26     629.79   
2                527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

In [187]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
data["line_cost"] = (data["unit_cost"] * data["qty"]) + data["shipping_price"]
data["line_cost"] = data["line_cost"].round(2)

data.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   subtotal_calculation  total_weight  shipping_price     line_price  \
0             115164.00        787.50         5512.50  131839.076250   
1                523.95         31.29          312.90     914.258625   
2                527.28         65.52          458.64    1077.117600   

   line_cost  line_profit  
0   85597.05     46242.03  
1     629.79       284.47  
2     765.18       311.94  

[3 rows x 24 columns]

In [182]:
# Create a column for the profit of
# each line using line cost and line price
data["line_profit"] = data["line_price"] - data["line_cost"]
data["line_profit"] = data["line_profit"].round(2)

data.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   subtotal_calculation  total_weight  shipping_price     line_price  \
0             115164.00        787.50         5512.50  131839.076250   
1                523.95         31.29          312.90     914.258625   
2                527.28         65.52          458.64    1077.117600   

   line_cost  line_profit  
0   85597.05     46242.03  
1     629.79       284.47  
2     765.18       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [192]:
#For loop, work on figuring out line_price
order_total_ids = { 
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}
order_totals = []
for order_id in order_total_ids:
    order_data = data[data["order_id"] == order_id]
    if not order_data.empty:
        given_total_price = order_total_ids[order_id]
        calculated_total_price = order_data["line_price"].sum()
        order_totals.append((order_id, given_total_price, calculated_total_price))
    else: print(f"Not Found in the Data")
for order_id, given_total_price, calculated_total_price in order_totals:
    if calculated_total_price == given_total_price:
        print(f"Order {order_id} total: ${given_total_price:.2f}")
    else:
        print(f"Order {order_id} total: ${calculated_total_price:.2f}")


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


In [193]:
# Check your work using the totals above
order_total_ids = { 
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}
order_totals = []
for order_id in order_total_ids:
    order_data = data[data["order_id"] == order_id]
    if not order_data.empty:
        given_total_price = order_total_ids[order_id]
        calculated_total_price = order_data["line_price"].sum()
        order_totals.append((order_id, given_total_price, calculated_total_price))
    else: print(f"Not Found in the Data")
for order_id, given_total_price, calculated_total_price in order_totals:
    if calculated_total_price == given_total_price:
        print(f"Order {order_id} total: ${given_total_price:.2f}")
    else:
        print(f"Order {order_id} total: ${calculated_total_price:.2f}")


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [194]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_5_client_entries = data[data["client_id"].isin(top_5_client_ids)].copy()
client_expenditures = top_5_client_entries.groupby("client_id")["line_price"].sum()
for client_id, total_spending in client_expenditures.items():
    print(f"{client_id}: {total_spending:.2f}")

24741: 82268892.02
33615: 8377308.52
38378: 12906550.87
46820: 9743794.36
66037: 10259514.79


In [226]:
top_5_client_entries = data[data["client_id"].isin(top_5_client_ids)].copy()
summary_df = top_5_client_entries.groupby("client_id").agg(
    qty=("qty","sum"),
    shipping_price=("shipping_price", "sum"),
    line_price=("line_price", "sum"),
    line_cost=("line_cost", "sum")
)

summary_df["line_profit"] = summary_df["line_price"] - summary_df["line_cost"]

summary_df.reset_index(inplace=True)

summary_df = summary_df.sort_values(by="line_profit", ascending=False)

summary_df.head()


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [232]:
top_5_client_entries = data[data["client_id"].isin(top_5_client_ids)].copy()
summary_df = top_5_client_entries.groupby("client_id").agg(
    qty=("qty","sum"),
    shipping_price=("shipping_price", "sum"),
    line_price=("line_price", "sum"),
    line_cost=("line_cost", "sum")
)
summary_df["line_profit"] = summary_df["line_price"] - summary_df["line_cost"]

summary_df.reset_index(inplace=True)

summary_df = summary_df.sort_values(by="line_profit", ascending=False)

summary_df.head()

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.


client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [271]:
top_5_client_entries = data[data["client_id"].isin(top_5_client_ids)].copy()
summary_df = top_5_client_entries.groupby("client_id").agg(
    qty=("qty","sum"),
    shipping_price=("shipping_price", "sum"),
    line_price=("line_price", "sum"),
    line_cost=("line_cost", "sum")
)
summary_df["Total Revenue Holder"] = summary_df["line_price"]
summary_df["Total Cost Holder"] = summary_df["line_cost"]
summary_df["Total Profit Holder"] = summary_df["line_price"] - summary_df["line_cost"]
summary_df["Shipping"] = summary_df["shipping_price"]

summary_df.reset_index(inplace=True)

summary_df["Total Revenue"] = (summary_df["Total Revenue Holder"] / 1000000).map("${:.2f}M".format)
summary_df["Total Cost"] = (summary_df["Total Cost Holder"] / 1000000).map("${:.2f}M".format)
summary_df["Total Profit"] = (summary_df["Total Profit Holder"] / 1000000).map("${:.2f}M".format)
summary_df["Shipping"] = (summary_df["shipping_price"] / 1000000).map("${:.2f}M".format)

summary_df.rename(columns={
    "client_id": "Client ID",
    "qty": "Units",
    "shipping_price": "Shipping",
    "line_price": "Total Revenue",
    "line_cost": "Total Cost",
    "line_profit": "Total Profit"
}, inplace=True)

summary_df = summary_df.sort_values(by="Total Profit", ascending=False)

summary_df.drop(columns=["Total Revenue Holder", "Total Cost Holder", "Total Profit Holder"], inplace=True)

summary_df.head()

Client ID   Units    Shipping Total Revenue   Total Cost Shipping  \
0      24741  239862  5126448.37   82268892.02  45688899.71   $5.13M   
2      38378   73667  3429455.40   12906550.87   9634720.98   $3.43M   
4      66037   43018  1395151.85   10259514.79   7004482.98   $1.40M   
3      46820   75768  1601448.84    9743794.36   7007191.64   $1.60M   
1      33615   64313  1828984.89    8377308.52   6175313.91   $1.83M   

  Total Revenue Total Cost Total Profit  
0       $82.27M    $45.69M      $36.58M  
2       $12.91M     $9.63M       $3.27M  
4       $10.26M     $7.00M       $3.26M  
3        $9.74M     $7.01M       $2.74M  
1        $8.38M     $6.18M       $2.20M

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.




In [273]:
top_5_client_entries = data[data["client_id"].isin(top_5_client_ids)].copy()
summary_df = top_5_client_entries.groupby("client_id").agg(
    qty=("qty","sum"),
    shipping_price=("shipping_price", "sum"),
    line_price=("line_price", "sum"),
    line_cost=("line_cost", "sum")
)
summary_df["Total Revenue Holder"] = summary_df["line_price"]
summary_df["Total Cost Holder"] = summary_df["line_cost"]
summary_df["Total Profit Holder"] = summary_df["line_price"] - summary_df["line_cost"]
summary_df["Shipping"] = summary_df["shipping_price"]

summary_df.reset_index(inplace=True)

summary_df["Total Revenue"] = (summary_df["Total Revenue Holder"] / 1000000).map("${:.2f}M".format)
summary_df["Total Cost"] = (summary_df["Total Cost Holder"] / 1000000).map("${:.2f}M".format)
summary_df["Total Profit"] = (summary_df["Total Profit Holder"] / 1000000).map("${:.2f}M".format)
summary_df["Shipping"] = (summary_df["shipping_price"] / 1000000).map("${:.2f}M".format)

summary_df.rename(columns={
    "client_id": "Client ID",
    "qty": "Units",
    "shipping_price": "Shipping",
    "line_price": "Total Revenue",
    "line_cost": "Total Cost",
    "line_profit": "Total Profit"
}, inplace=True)

summary_df = summary_df.sort_values(by="Total Profit", ascending=False)

summary_df.drop(columns=["Total Revenue Holder", "Total Cost Holder", "Total Profit Holder"], inplace=True)

summary_df.head()

Client ID   Units    Shipping Total Revenue   Total Cost Shipping  \
0      24741  239862  5126448.37   82268892.02  45688899.71   $5.13M   
2      38378   73667  3429455.40   12906550.87   9634720.98   $3.43M   
4      66037   43018  1395151.85   10259514.79   7004482.98   $1.40M   
3      46820   75768  1601448.84    9743794.36   7007191.64   $1.60M   
1      33615   64313  1828984.89    8377308.52   6175313.91   $1.83M   

  Total Revenue Total Cost Total Profit  
0       $82.27M    $45.69M      $36.58M  
2       $12.91M     $9.63M       $3.27M  
4       $10.26M     $7.00M       $3.26M  
3        $9.74M     $7.01M       $2.74M  
1        $8.38M     $6.18M       $2.20M

In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M